pkg/util/fs.go

```diff
@@ -32,34 +35,111 @@ type FileSystem interface {
 	WriteFile(file string, data []byte) error
 	ReadDir(string) ([]os.FileInfo, error)
 	Stat(string) (os.FileInfo, error)
+	Walk(string, filepath.WalkFunc) error
 }
 
 // NewFileSystem creates a new instance of the default FileSystem
 // implementation
 func NewFileSystem() FileSystem {
 	return &fs{
-		runner: NewCommandRunner(),
+		runner:    NewCommandRunner(),
+		fileModes: make(map[string]os.FileMode),
 	}
 }
 
 type fs struct {
-	runner CommandRunner
+	runner    CommandRunner
+	fileModes map[string]os.FileMode
+	m         sync.Mutex
 }
 
+// FileInfo is a struct which implements os.FileInfo.  We use it (a) for test
+// purposes, and (b) because we enrich the FileMode on Windows systems
+type FileInfo struct {
+	FileName    string
+	FileSize    int64
+	FileMode    os.FileMode
+	FileModTime time.Time
+	FileIsDir   bool
+	FileSys     interface{}
+}
+
+// Name retuns the filename of fi
+func (fi *FileInfo) Name() string {
+	return fi.FileName
+}
+
+// Size returns the file size of fi
+func (fi *FileInfo) Size() int64 {
+	return fi.FileSize
+}
+
+// Mode returns the file mode of fi
+func (fi *FileInfo) Mode() os.FileMode {
+	return fi.FileMode
+}
+
+// ModTime returns the file modification time of fi
+func (fi *FileInfo) ModTime() time.Time {
+	return fi.FileModTime
+}
+
+// IsDir returns true if fi refers to a directory
+func (fi *FileInfo) IsDir() bool {
+	return fi.FileIsDir
+}
+
+// Sys returns the sys interface of fi
+func (fi *FileInfo) Sys() interface{} {
+	return fi.FileSys
+}
+
+func copyFileInfo(src os.FileInfo) *FileInfo {
+	return &FileInfo{
+		FileName:    src.Name(),
+		FileSize:    src.Size(),
+		FileMode:    src.Mode(),
+		FileModTime: src.ModTime(),
+		FileIsDir:   src.IsDir(),
+		FileSys:     src.Sys(),
+	}
+}
+
+// Stat returns a FileInfo describing the named file.
 func (h *fs) Stat(path string) (os.FileInfo, error) {
-	return os.Stat(path)
+	fi, err := os.Stat(path)
+	if runtime.GOOS == "windows" && err == nil {
+		h.m.Lock()
+		fi = h.enrichFileInfo(path, fi)
+		h.m.Unlock()
+	}
+	return fi, err
 }
 
+// ReadDir reads the directory named by dirname and returns a list of directory
+// entries sorted by filename.
 func (h *fs) ReadDir(path string) ([]os.FileInfo, error) {
-	return ioutil.ReadDir(path)
+	fis, err := ioutil.ReadDir(path)
+	if runtime.GOOS == "windows" && err == nil {
+		h.m.Lock()
+		for i := range fis {
```

Comment:
Thanks for the feedback on this - next push will include documentation and improved locking along the lines of what you suggested.